design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,290 @@
+
+There are two steps in the rebalancing procedure that are expensive: 
+  1) One of `B` or `C` is picked as the new active, this takes ~10 seconds based on default configs. 
+  2) Once a new active is chosen, depending on how far behind (lag) it is with respect to the changelog, 
+  it may take up  few seconds or even minutes before the standby has fully caught up to the last committed offset. 
+
+In total, it takes >>10 seconds before a query can start succeeding again. What can we do until a
+new active is ready to serve requests? 
+[KIP-535](https://cwiki.apache.org/confluence/display/KAFKA/KIP-535%3A+Allow+state+stores+to+serve+stale+reads+during+rebalance) 
+has laid the groundwork to allow us to serve requests from standbys, even if they are still in 
+rebalancing state or have not caught up to the last committed offset of the active. 
+
+Every ksql server now has the information of:
+
+1. Local offset lag: The lag of all partitions the server hosts (whether active or standby)
+2. Routing Metadata for a key: The partition, active host and standy hosts per key
+
+This information allows each server (with some communication, discussed below) to know the global 
+lag information of every topic partition whether it is the standby or active. This enables us to 
+implement a policy where `B` having established that `A` is down (after trying to send a request 
+to `A` and seeing it failed), to decide whether it (`B`) will serve it or `C`. This effectively 
+means that there will be little to no down time in serving requests at the cost of consistency as
+`B` or `C` may serve stale (not caught up) data. Hence, we achieve high availability at the cost of 
+consistency. Eventual consistency for the win!
+
+
+## What is in scope
+Ensure availability for pull queries when KSQL servers fail (with `ksql.streams.num.standby.replicas=N`, 
+we can tolerate upto N such failures) 
+
+## What is not in scope
+
+- Address unavailability caused by cold starts of KSQL server i.e when a new server is added to a 
+KSQL cluster, it must first rebuild its state off the changelog topics and that process still could 
+take a long time. 
+- Try to improve consistency guarantees provided by KSQL i.e reduce the amount of time it takes to 
+rebalance or standby replication to catch up. 
+
+## Value/Return
+
+The cluster of KSQl server will be able to serve requests even when the active is down. This mitigates 
+a large current gap in deploying KSQL pull queries for mission critical use-cases, by significantly 
+reducing failure rate of pull queries during server failures.
+
+## Public APIS
+No change to public APIs
+
+## Design
+
+The problem at hand is how to determine if a server is down and where to route a request when it is 
+down. Every KSQL server must have available the information of what other KSQL servers exist in the 
+cluster, their status and their lag per topic partition. There are three components to this: 
```

Comment:
Yes, we will be periodically calling the allMetadata method. I am not sure if there is a way to add a callback.

 Regarding the problem if there are exist no persistent queries, then no routing will be needed either :) From our offline discussion, as soon as a server joins, it reads from the command topic and executes all persistent queries. Hence, every server in the cluster will know about all persistent queries whether it gets assigned a partition from the rebalance process or not. (We might want to verify if this actually the case but looking at the code, it seems to be true)

Reply:
Keep in mind headless mode with no command topic! (Thought there is a config topic we could use).

I don't think we should be using `allMetadata` to determine cluster membership.  This is a per-query bit of metadata.  You've already identified the no-runnning-queries edge case, but also consider that in the future we may not have all queries running on all nodes. (as that sucks for scalability).

We should decouple cluster membership from where queries are running. I would suggest wrapping the consumer group protocol around a known topic, (command/config), in a class that provides callbacks for nodes leaving / joining.  This consumer group can then be tuned independently of any query's consumer group.

Reply:
When queries no longer spread evenly across nodes, we need to also gossip/exchange routing information somehow. I would like to keep this tied to StreamsMetadata for now. This can be changed down the line.

Reply:
Just to confirm my understanding, `allMetadata` is basically iterating through all streams applications on a server, extracting their `StreamsMetadata`, and then doing a union of all the hosts / ports to generate the list of Ksql servers?

Reply:
@apurvam  yes, this is what we are doing.

Reply:
Tracked in issue #4228